server/src/main/java/io/crate/execution/engine/collect/sources/InformationSchemaIterables.java

```diff
@@ -414,6 +422,60 @@ private void createMetadataBasedIterables(Metadata metadata) {
         routines = () -> sequentialStream(routineInfos).filter(Objects::nonNull).iterator();
     }
 
+    public Set<String> enabledRoles(Role role, Roles roles) {
+        Set<String> result = new HashSet<>();
+        Set<ApplicableRole> applicableRoles = applicableRoles(role, roles);
+        for (ApplicableRole applicableRole: applicableRoles) {
+            result.add(applicableRole.roleName());
+        }
+        return result;
+    }
+
+    public Set<ApplicableRole> applicableRoles(Role role, Roles roles) {
+        Set<ApplicableRole> applicableRoles = new HashSet<>();
+
+        for (GrantedRole grantedRole: role.grantedRoles()) {
+            ApplicableRole newRole = new ApplicableRole(
+                role.name(),
+                grantedRole.roleName(),
+                roles.hasPrivilege(role, Permission.AL, Securable.CLUSTER, null)
+            );
+            Role retrievedRole = roles.findRole(grantedRole.roleName());
+            if (retrievedRole != null) {
+                if (!retrievedRole.grantedRoles().isEmpty()) {
+                    applicableRoles.addAll(applicableRoles(retrievedRole, roles));
+                }
+            }
+            applicableRoles.add(newRole);
+        }
+
+        return applicableRoles;
+    }
```

Comment:
To avoid allocations, we could use the stream api to return iterables:
```suggestion
    public Iterable<ApplicableRole> applicableRoles(Role role, Roles roles) {
        return () -> {
            Stream<ApplicableRole> s = role.grantedRoles().stream()
                .mapMulti((grantedRole, c) -> {
                    c.accept(new ApplicableRole(grantedRole.grantor(), grantedRole.roleName(), roles.hasPrivilege(role, Permission.AL, Securable.CLUSTER, null)));
                    Role retrievedRole = roles.findRole(grantedRole.roleName());
                    if (retrievedRole != null) {
                        retrievedRole.grantedRoles().stream()
                            .map(gr -> new ApplicableRole(grantedRole.grantor(), grantedRole.roleName(), roles.hasPrivilege(role, Permission.AL, Securable.CLUSTER, null)))
                            .forEach(c);
                    }
                });
            return s.iterator();
        };
    }
```

Additionally, I think there was a mistake, as `role.name()` was passed into `ApplicableRole` as the `grantee`?